File: Civitas ruby/lib/jugador.rb
#encoding: utf-8

module Civitas
class Jugador
  @@CASASMAX = 4
  @@CASASPORHOTEL = 4
  @@HOTELESMAX = 4
  @@PASOPORSALIDA = 1000
  @@PRECIOLIBERTAD = 200
  @@SALDOINICIAL = 7500 
  
  def initialize(propiedades, encarcelado, nombre, numCasillaActual, puedeComprar, salvoconducto)
    @propiedades = propiedades
    @encarcelado = encarcelado
    @nombre = nombre
    @numCasillaActual = numCasillaActual
    @puedeComprar = puedeComprar
    @saldo = @@SALDOINICIAL
    @salvoconducto = salvoconducto
  end
  
  def cancelarHipoteca(ip)
    result = false;
    
    if(@encarcelado)
      return result
    end
    
    if(existeLaPropiedad(ip))
      propiedad = @propiedades[ip]
      cantidad = propiedad.getImporteCancelarHipoteca
      puedoGastar = puedoGastar(cantidad)
      
      if(puedoGastar)
        result = propiedad.cancelarHipoteca
        result = false
        
        if(propiedad.getHipotecado)
          if(propiedad.esEsteElPropietario(self))
            paga(propiedad.getImporteCancelarHipoteca)
            propiedad.hipotecar
            result = true
          end
        end
        
        if(result)
          Diario.instance.ocurre_evento("El jugador " + @nombre + " cancela la hipoteca de la propiedad " + ip.to_s)
        end
      end
    end
    
    return result
  end
  
  def cantidadCasasHoteles
    cantidad = 0
    
    for propiedad in @propiedades
      cantidad = cantidad + propiedad.cantidadCasasHoteles;
    end
    
    return cantidad
  end
  
  def comprar(titulo)
    result = false
    
    if(@encarcelado)
      return result
    end
    
    if(@puedeComprar)
      precio = titulo.getPrecioCompra
      
      if(puedoGastar(precio))
        result = titulo.comprar(self)
        result = false
        
        if(!titulo.tienePropietario)
          titulo.actualizaPropietarioPorConversion(self)
          result = true
          paga(precio)
        end
        
        if(result)
          @propiedades << titulo
          Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " compra la propiedad " + titulo.getNombre)
        end
        @puedeComprar = false
      end
    end
    
    return result
  end
  
  
  #EL JUGADOR NO PAGA POR EDIFICAR?
  def construirCasa(ip)
    result = false
    puedoEdificarCasa = false
    
    if(@encarcelado)
      return result
    end
    
    if(!@encarcelado)
      existe = existeLaPropiedad(ip)
      
      if(existe)
        propiedad = @propiedades[ip]
        puedoEdificarCasa = puedoEdificarCasa(propiedad)
        precio = propiedad.getPrecioEdificar
        
        if(puedoGastar(precio) && @@CASASMAX > propiedad.getNumCasas)
          puedoEdificarCasa = true
        end
        
        if(puedoEdificarCasa)
          result = propiedad.construirCasa(ip)
          paga(precio)
          
          if(result)
            Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " construye casa en la propiedad " + ip.to_s)
          end
        end
      end
    end
    
    return result
  end
  
  def construirHotel(ip)
    result = false
    
    if(@encarcelado)
      return result
    end
    
    if(existeLaPropiedad(ip))
      propiedad = @propiedades[ip]
      puedoEdificarHotel = puedoEdificarHotel(propiedad)
      puedoEdificarHotel = false
      precio = propiedad.getPrecioEdificar
      
      if(puedoGastar(precio))
        if(propiedad.getNumHoteles < @@HOTELESMAX)
          if(propiedad.getNumCasas >= @@CASASPORHOTEL)
            puedoEdificarHotel = true
          end
        end
      end
      
      if(puedoEdificarHotel)
        result = propiedad.construirHotel(ip)
        result = false
        
        if(propiedad.esEsteElPropietario(self))
          paga(precio)
          propiedad.construirHotel(ip)
          result = true
        end
        
        casasPorHotel = getCasasPorHotel
        propiedad.derruirCasas(casasPorHotel, self)
        Diario.instance().ocurre_evento("\nEl jugador " + @nombre + " construye un hotel en la propiedad " + ip.to_s)
      end
    end
    
    return result
  end
  
  def enBancarrota
    return @saldo < 0
  end
  
  
  def encarcelar(numCasillaCarcel)
    if(debeSerEncarcelado())
      moverACasilla(numCasillaCarcel)
      @encarcelado = true
      Diario.instance.ocurre_evento("\nEl jugador: " + @nombre + " entra a la carcel")
    end
    return @encarcelado
  end
  
  def getCasasPorHotel
    return @@CASASPORHOTEL
  end
  
  def getNumCasillaActual
    return @numCasillaActual
  end
  
  def getPuedeComprar
    return @puedeComprar
  end
  
  def hipotecar(ip)
    result = false
    
    if(@encarcelado)
      return result
    end
    
    if(existeLaPropiedad(ip))
      propiedad = @propiedades[ip]
      result = propiedad.hipotecar(self)
      salida = false
      
      if(!propiedad.getHipotecado && propiedad.esEsteElPropietario(self))
        recibe(propiedad.getImporteCancelarHipoteca)
        propiedad.hipotecar(self)
        salida = true
      end
      
      result = salida
    end
    
    if(result)
      Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " hipoteca la propiedad " + ip.to_s)
    end
    
    return result
  end
  
  def self.nombre(nombre)
    self.new([], false, nombre, 0, true, false)
  end
  
  def self.copia(otro)
    @nombre = otro.nombre 
    @propiedades = otro.propiedades
    @encarcelado = otro.encarcelado
    @numCasillaActual = otro.numCasillaActual
    @puedeComprar = otro.puedeComprar
    @saldo = otro.saldo
    @salvoconducto = otro.salvoconducto
    return self
  end
  
  def modificarSaldo(cantidad)
    @saldo = @saldo + cantidad
    Diario.instance.ocurre_evento("\nSaldo de " + @nombre + " modificado a: " + @saldo.to_s)
  end
  
  def moverACasilla(numCasilla)
    if(@encarcelado)
      return false
    else
      @numCasillaActual = numCasilla
      @Comprar = false
      Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " se mueve a la casilla " + numCasilla.to_s)
      
      return true
    end
  end
  
  def obtenerSalvoConducto(sorpresa)
    if(@encarcelado)
      return false
    else
      @salvoconducto = sorpresa
    end
  end
  
  def paga(cantidad)
    valor = modificarSaldo(cantidad * -1)
    
    return valor
  end
  
  def pagaAlquiler(cantidad)
    if(@encarcelado)
      return false
    else
      valor = paga(cantidad)
      return valor
    end
  end
  
  def pagaImpuesto(cantidad)
    if(@encarcelado)
      return false
    else
      valor = paga(cantidad)
      return valor
    end
  end
  
  def pasaPorSalida
    modificarSaldo(@@PASOPORSALIDA)
    Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " pasa por salida")
    return true
  end
  
  def puedeComprarCasilla
    if(@encarcelado)
      @puedeComprar = false
    else
      @puedeComprar = true
    end
    
    return @puedeComprar
  end
  
  def recibe(cantidad)
    if(@encarcelado)
      return false
    else
      valor = modificarSaldo(cantidad)
      return valor
    end
  end
  
  def salirCarcelPagando
    if(@encarcelado && puedeSalirCarcelPagando())
      paga(@@PRECIOLIBERTAD)
      @encarcelado = false
      Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " sale de la carcel pagando")
      return true
    else
      return false
    end
  end
  
  def salirCarcelTirando
    resultado = Dado.salgoDeLaCarcel()
    if(resultado)
      @encarcelado = false
      Diario.instance.ocurre_evento("\nEl jugador " + @nombre + " sale de la carcel tirando")
    end
    
    return @encarcelado
  end
  
  def tieneAlgoQueGestionar
    return @propiedades.length > 0
  end
  
  def tieneSalvoConducto
    if(@salvoconducto == nil)
      return true
    else
      return false
    end
  end
  
  
  
  
  
  
  
  #El titulo de propiedad con numero ip es asi?
  def vender(ip)
    if(@encarcelado)
      return false
    elsif (existeLaPropiedad(ip))
      resultado = @propiedades[ip].vender(self)
      if(resultado)
        @propiedades.delete_at(ip)
        Diario.instance.ocurre_evento("\nSe vende la propiedad")
        return true 
      else
        return false
      end
    else
      return false
    end
  end
  
  private
  
  def existeLaPropiedad(ip)
    return @propiedades.length > ip
  end
  
  def getCasasMax()
    return @@CASASMAX
  end
  
  def getHotelesMax
    return @@HOTELESMAX
  end
  
  def getPrecioLibertad
    return @@PRECIOLIBERTAD
  end  
    
  def getPremioPasoSalida
    return @@PASOPORSALIDA
  end  
    
  def perderSalvoConducto
    @salvoconducto.usada()
    @salvoconducto = nil
  end
  
  def puedeSalirCarcelPagando
    if(@saldo >= @@PRECIOLIBERTAD)
      return true
    else
      return false
    end
  end
      
  def puedoEdificarCasa(propiedad)
    
  end
  
  def puedoEdificarHotel(propiedad)
    
  end
  
  def puedoGastar(precio)
    if(@encarcelado)
      return false
    elsif (@saldo >= precio)
      return true
    else
      return false
    end
  end
  
  public
  
  def <=>(otro)
    return @saldo <=> otro.getSaldo
  end  
    
  def isEncarcelado
    if(@encarcelado)
      return true
    else
      return false
    end
  end
  
  def toString
    cadena = "\nJugador: " + @nombre.to_s +
      "\nSaldo: " + @saldo.to_s + 
      "\nEncarcelado: " + @encarcelado.to_s +
      "\nSalvoconducto: " +@salvoconducto.to_s;
    
    return cadena;
        
  end
  
    
  def debeSerEncarcelado
    if(@encarcelado)
      return true
    elsif (!tieneSalvoConducto())
      return true
    elsif (tieneSalvoConducto())
      perderSalvoConducto()
      Diario.instance.ocurre_evento("\nEl jugador se libra de la cárcel con salvoconducto")
    end
  end
  
  def getNombre
    return @nombre
  end
  
  def getPropiedades
    return @propiedades
  end  
    
  def getSaldo
    return @saldo
  end
  
  def recibeCasa(propiedad)
    if(!@encarcelado)
      @propiedades << propiedad
      return true
    else
      return false
    end
  end
  
  def getPropiedades
    return @propiedades
  end
  
  def eliminaCasa(ip)
    @propiedades.delete_at(ip)
  end
  
end
end

File: Civitas java/src/civitas/Jugador.java
package civitas;

import java.util.ArrayList;

public class Jugador implements Comparable<Jugador> {
    protected static final int CasasMax = 4;
    protected static final int CasasPorHotel = 4;
    protected boolean encarcelado;
    protected static final int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    protected static final float PasoPorSalida = 1000;
    protected static final float PrecioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    private static final float SaldoInicial = 7500;
    private ArrayList<TituloPropiedad> propiedades;
    private Sorpresa Salvoconducto;
    
    
    //propiedad.hipotecar pone hipotecado a true?
    boolean cancelarHipoteca(int ip){
       boolean result = false;
       
       if(encarcelado)
           return result;
       else if(existeLaPropiedad(ip)){
           TituloPropiedad propiedad = propiedades.get(ip);
           float cantidad = propiedad.getImporteCancelarHipoteca();
           boolean puedoGastar = puedoGastar(cantidad);
           
           if(puedoGastar){
               result = propiedad.cancelarHipoteca(this);
               if(!result){
                   if(propiedad.getHipotecado()){
                       if(propiedad.esEsteElPropietario(this)){
                           paga(propiedad.getImporteCancelarHipoteca());
                           propiedad.hipotecar(this);
                           result = true;
                       }
                   }
                       
               } else
                   Diario.getInstance().ocurreEvento("\nEl jugador "+nombre+ " cancela la hipoteca de la propiedad "+ip);
           }
       }
       return result;
    }
    
    
    int cantidadCasasHoteles(){
        int cantidad = 0;
        
        for(int i = 0; i < propiedades.size(); i++)
            cantidad += propiedades.get(i).cantidadCasasHoteles();
        
        return cantidad;
    }
    
    public int compareTo(Jugador otro){
        return Float.compare(saldo, otro.saldo);
    }
    
    boolean comprar(TituloPropiedad titulo){
        boolean result = false;
        
        if(encarcelado){
            return result;
        } else if(puedeComprar){
            float precio = titulo.getPrecioCompra();
            
            if(puedoGastar(precio)){
                result = titulo.comprar(this);
                result = false;
                
                if(!titulo.tienePropietario()){
                    titulo.actualizaPropietarioPorConversion(this);;
                    result = true;
                    paga(titulo.getPrecioCompra());
                }
                
                if(result){
                    propiedades.add(titulo);
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + " compra la propiedad " + titulo.toString());
                }
                puedeComprar = false;
            }
        }
        return result;
    }
    
    boolean construirCasa(int ip){
        boolean result = false;
        boolean puedoEdificarCasa = false;
        TituloPropiedad propiedad = null;
        
        if(encarcelado)
            return result;
        else{
            boolean existe = existeLaPropiedad(ip);
            
            if(existe)
                propiedad = propiedades.get(ip);
                puedoEdificarCasa = puedoEdificarCasa(propiedad);
                float precio = propiedad.getPrecioEdificar();
                
                if(puedoGastar(precio)){
                    if(propiedad.getNumCasas() < getCasasMax())
                        puedoEdificarCasa = true;
                }
                
                if(puedoEdificarCasa)
                    result = propiedad.construirCasa(this);
                
                if(result)
                    Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " construye una casa en la propiedad " + ip);
        }
        return result;  
    }
    
    //propiedad.constuirHotel para incrementar el numero de hoteles?
    boolean construirHotel(int ip){
        boolean result = false;
        
        if(encarcelado){
            return result;
        } else if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            boolean puedoEdificarHotel = puedoEdificarHotel(propiedad);
            puedoEdificarHotel = false;
            float precio = propiedad.getPrecioEdificar();
            
            if(puedoGastar(precio)){
                if(propiedad.getNumHoteles() < getHotelesMax()){
                    if(propiedad.getNumCasas() >= getCasasPorHotel())
                        puedoEdificarHotel = true;
                }
            }
            
            if(puedoEdificarHotel){
                result = propiedad.construirHotel(this);
                result = false;
                
                if(propiedad.esEsteElPropietario(this)){
                    paga(propiedad.getPrecioEdificar());
                    propiedad.construirHotel(this);
                    result = true;
                }
                
                int casasPorHotel = getCasasPorHotel();
                propiedad.derruirCasas(casasPorHotel, this);
                Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " construye hotel en la propiedad " + ip);
            }
        }
        return result;
    }
    
    protected boolean debeSerEncarcelado(){
        if(isEncarcelado())
           return false;
        else if(!tieneSalvoconducto())
            return true;
        else if(tieneSalvoconducto()){
            perderSalvoconducto();
            Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " pierde el salvoconducto y se libra de la cárcel");
            return false;
        } else
            return false;
    }
    
    boolean enBancarrota(){
        if(saldo < 0)
            return true;
        else
            return false;
    }

    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " es encarcelado");
        }
        return encarcelado;
    }
    
    private boolean existeLaPropiedad(int ip){
        return ip < propiedades.size();
    }
    
    private int getCasasMax(){
        return CasasMax;
    }
    
    private int getHotelesMax(){
        return HotelesMax;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    protected String getNombre(){
        return nombre;
    }
    
    public int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
    
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    protected float getSaldo(){
        return saldo;
    }
    
    //propiedad.hipotecar para poner hipotecado a true?
    boolean hipotecar(int ip){
        boolean result = false;
        
        if(encarcelado){
            return result;
        }
        
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            result = propiedad.hipotecar(this);
            boolean salida = false;
            
            if(!propiedad.getHipotecado() && propiedad.esEsteElPropietario(this)){
                recibe(propiedad.getImporteHipoteca());
                propiedad.hipotecar(this);
                salida = true;
            }
            result = salida;
        }
        
        if(result){
            Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " hipoteca la propiedad " + ip);
        }
        
        return result;
    }
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    Jugador(String nombre1){
        propiedades = new ArrayList<TituloPropiedad>();
        nombre = nombre1;
        numCasillaActual = 0;
        encarcelado = false;
        saldo = SaldoInicial;
        puedeComprar = true;
        Salvoconducto = null;
    }
    
    protected Jugador(Jugador otro){
        propiedades = otro.propiedades;
        numCasillaActual = otro.numCasillaActual;
        nombre = otro.nombre;
        encarcelado = otro.encarcelado;
        saldo = otro.saldo;
        puedeComprar = otro.puedeComprar;
        Salvoconducto = otro.Salvoconducto;
    }
    
    boolean modificarSaldo(float cantidad){
        saldo += cantidad;
        Diario.getInstance().ocurreEvento("\nSe modifica el saldo del jugador " + nombre + " a " + saldo);
        return true;    
    }
    
    boolean moverACasilla(int numCasilla){
        if(isEncarcelado())
            return false;
        else {
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " se mueve a la casilla " + numCasillaActual);
            return true;
        }
    }
    
    boolean obtenerSalvoConducto(Sorpresa sorpresa){
        if(isEncarcelado())
            return false;
        else {
            Salvoconducto = sorpresa;
            return true;
        }
        
    }
    
    boolean paga(float cantidad){
        return modificarSaldo(cantidad * -1);
    }
    
    boolean pagaAlquiler(float cantidad){
        if(isEncarcelado())
            return false;
        else
            return paga(cantidad);

    }
    
    boolean pagaImpuesto(float cantidad){
        if(isEncarcelado())
            return false;
        else
            return paga(cantidad);

    }
    
    boolean pasoPorSalida(){
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " pasa por salida");
        return true;
    }
    
    private void perderSalvoconducto(){
        Salvoconducto.usada();
        Salvoconducto = null;
    }
    
    boolean puedeComprarCasilla(){
        if(isEncarcelado())
            puedeComprar = false;
        else
            puedeComprar = true;
        
        return puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando(){
        return saldo >= PrecioLibertad;
    }
    
    
    //Completar adecuadamente
    private boolean puedoEdificarCasa(TituloPropiedad titulo){
        if(titulo.getNumCasas() < CasasMax && titulo.getPrecioEdificar() <= saldo)
            return true;
        else
            return false;
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad titulo){
        if(titulo.getNumHoteles() < HotelesMax && titulo.getPrecioEdificar() <= saldo)
            return true;
        else
            return false;
    }
    
    private boolean puedoGastar(float cantidad){
        if(isEncarcelado())
            return false;
        else
            return saldo >= cantidad;
    }
    
    boolean recibe(float cantidad){
        if(isEncarcelado())
            return false;
        else
            return modificarSaldo(cantidad);
    }
    
    boolean salirCarcelPagando(){
        if(isEncarcelado() && puedeSalirCarcelPagando()){
            paga(PrecioLibertad);
            encarcelado = false;
            Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " sale de la carcel pagando");
            return true;
        } else
            return false;
    }
    
    boolean salirCarcelTirando(){
        if(isEncarcelado()){
            if(Dado.getInstance().salgoDeLaCarcel()){
                encarcelado = false;
                Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " sale de la carcel tirando");
            }
        }
        return encarcelado;
    }
    
    boolean tieneAlgoQueGestionar(){
        return propiedades.size() > 0;
    }
    
    boolean tieneSalvoconducto(){
        if(Salvoconducto != null)
            return true;
        else
            return false;
    }
    
    public String toString(){
        String cadena = "\nJugador: " + nombre +
                "\nSaldo: " + saldo + 
                "\nEncarcelado: " + encarcelado +
                "\nSalvoconducto: " + Salvoconducto;
        
        return cadena;
    }
    
    boolean vender(int ip){
        boolean resultado;
        if(isEncarcelado())
            return false;
        else if(existeLaPropiedad(ip)){
            resultado = propiedades.get(ip).vender(this);
            if(resultado){
                propiedades.remove(ip);
                System.out.println("\n\nTAMAÑO PROPIEDADES: " + propiedades.size());
                Diario.getInstance().ocurreEvento("\nSe vende la propiedad");
                return true;
            } else
                return false;
        } else
            return false;
    }
    
    boolean recibeCasa(TituloPropiedad propiedad){
        if(!encarcelado){
            propiedades.add(propiedad);
            return true;
        } else
            return false;
    }
}
